### ISSUE TITLE
ORecordLazyList.get(index) started to return ORecordId instead of ODocument

### ISSUE THREAD

**Issue by PhantomYdn** [orientechnologies/orientdb#9421]
### OS: Windows, Linux

Some of our tests started to fail after switching from 3.1.3 to 3.1.4 due to ClassCastException.
After review we have figured out that problem is in ORecordLazyList: first item is ODocument, but all subsequent are just ORecordId (but a corresponding  ODocument does exist)

## Expected behavior  
ORecordLazyList should return ODocument

## Actual behavior  
All items after first one are ORecordId

## Steps to reproduce  
Not fully sure how to reproduce from scratch. The following our testcase stopped working: https://github.com/OrienteerBAP/Orienteer/blob/future/orienteer-notification/src/test/java/org/orienteer/notifications/test/TestNotificationLifecycle.java

**Comment by PhantomYdn** [comment]
@tglman `git bisect` between 3.1.3 and 3.1.4 show that the issue was introduced by the following commit: https://github.com/orientechnologies/orientdb/commit/6265642427def0a8c474c66e1619fe7b658c8815

**Comment by PhantomYdn** [comment]
@tglman, @laa, @luigidellaquila, OrientDB team, do you have ideas how to fix that? 3.1.4 is unusable due to this issue...

**Comment by PhantomYdn** [comment]
Apparently, the reason is in fact that mentioned above commit doesn't affect logic of `iterator()`. Defect is pretty visible in the following testcase:

```java
@Test
public void testLinkListAndODocumentConversion()
{
	ODatabaseSession db = wicket.getTester().getDatabaseSession();
	OSchemaHelper helper = OSchemaHelper.bind(db);
	helper.oClass("TestODocumentConverstion")
			.oProperty("name", OType.STRING)
			.oProperty("items", OType.LINKLIST);
	ODocument mainDoc = helper.oDocument("name", "document").saveDocument().getODocument();
	helper.oClass("TestODocumentConverstionItems")
		  	.oProperty("name", OType.STRING);
	ODocument doc1 = helper.oDocument("name", "item1").saveDocument().getODocument();
	ODocument doc2 = helper.oDocument("name", "item2").saveDocument().getODocument();
	ODocument doc3 = helper.oDocument("name", "item3").saveDocument().getODocument();
	List<ODocument> items = Arrays.asList(doc1, doc2, doc3);
	mainDoc.field("items", items);
	mainDoc.save();
	db.commit();
	db.begin();
	mainDoc = (ODocument)mainDoc.reload();
	items = mainDoc.field("items", List.class);
	items.iterator().next(); // IMPORTANT LINE
	assertTrue(items.get(0) instanceof ODocument);
	assertEquals(doc1, items.get(0));
	assertTrue(items.get(1) instanceof ODocument);
	assertEquals(doc2, items.get(1));
	assertTrue(items.get(2) instanceof ODocument);
	assertEquals(doc3, items.get(2));
	db.commit();
}
```
This test-case will fail, but if you comment out line marked as `IMPORTANT LINE` - test-case will be OK

**Comment by PhantomYdn** [comment]
Investigation shown that `HelperClasses.readLinkCollection()` should be able to adjust `ORecordLazyList.contentType`. Otherwise: during iterating method `OLazyRecordIterator.next()` invokes `update()` which switch `ORecordLazyList.contentType` from EMPTY to ALL_RECORDS

**Comment by PhantomYdn** [comment]
Issue was fixed by PR #9437 
Please merge

**Comment by PhantomYdn** [comment]
Issue can be closed, because corresponding PR was merged. Thank you, @tglman !
